watcher/src/block_data_store.rs

```diff
@@ -0,0 +1,339 @@
+// Copyright (c) 2018-2021 The MobileCoin Foundation
+
+//! An store object for managing storage of BlockData objects in the database, while taking care of
+//! de-duplicating contents when possible.
+
+use crate::error::WatcherDBError;
+use lmdb::{Cursor, Database, DatabaseFlags, Environment, RwTransaction, Transaction, WriteFlags};
+use mc_common::{
+    logger::{log, Logger},
+    HashMap,
+};
+use mc_crypto_digestible::{Digestible, MerlinTranscript};
+use mc_transaction_core::{Block, BlockContents, BlockData, BlockIndex, BlockSignature};
+use mc_util_serial::{decode, encode};
+use prost::Message;
+use std::{str::FromStr, sync::Arc};
+use url::Url;
+
+/// Block datas database name.
+pub const BLOCK_DATAS_BY_INDEX_DB_NAME: &str = "watcher_db:block_data:blocks_datas_by_index";
+
+/// Blocks by hash database name.
+pub const BLOCKS_BY_HASH_DB_NAME: &str = "watcher_db:block_data:blocks_by_hash";
+
+/// BlockContentss by hash database name.
+pub const BLOCK_CONTENTS_BY_HASH_DB_NAME: &str = "watcher_db:block_data:block_contents_by_hash";
+
+/// An internal object for representing BlockData that doesn't hold the actual Block and
+/// BlockContents since those might be shared with other blocks.
+#[derive(Clone, Message)]
+pub struct StoredBlockData {
+    /// 32 bytes hash of Block.
+    #[prost(bytes, required, tag = "1")]
+    pub block_hash: Vec<u8>,
+
+    /// 32 bytes hash of BlockContent.
+    #[prost(bytes, required, tag = "2")]
+    pub block_contents_hash: Vec<u8>,
+
+    /// Block signature (optional).
+    // The signature is unique (we do not expect to encounter duplicate signatures)
```

Comment:
What are the implications if we did encounter duplicate signatures?

Reply:
No implications, we'll store both of them. But its the reason we are storing the actual signature here and not a hash that points to a `BlockSignature` object (since we don't expect to ever have more than one this will not actually save space).